lib/mix/lib/mix/compilers/elixir.ex

```diff
@@ -227,7 +232,8 @@ defmodule Mix.Compilers.Elixir do
         runtime_references: runtime_references,
         compile_dispatches: compile_dispatches,
         runtime_dispatches: runtime_dispatches,
-        external: source_external
+        external: source_external,
+        warnings: source(source_entry, :warnings)
```

Comment:
I think it should never be empty, since all source files that are being compiled should be accounted for, but for safety I could change it to something like this:
```
      warnings = case source_entry do
        source(warnings: warnings) -> warnings
        nil -> []
      end
```